### DISCUSSION TITLE
PCGNS domain decomposition and zone connectivity questions

### DISCUSSION THREAD

**Discussion by papis84** [CGNS/CGNS#369] [CGNStalk]
Greetings,

We are trying to incorporate (P)CGNS in our unstructured CFD code. CGNS input is quite straightforward, nevertheless, writing proved a little bit trickier. The idea is that we have a big grid consinsting of N nodes and C cells. Using METIS we split the grid in the various processors so each one has [1,n] nodes and [1,c] elements. 

In order to write the grid output there are two approaches. The first is to define one zone consisting of N nodes and C cells and have each processor write each own data accumulatively. Unfortunately, since each processors uses a local indexing there is no connectivity between the processor domains (which might be problematic for producing isorsurfaces etc).

There other way around is to  create one zone for each processor write the nodes elements using the local indexing of each domain and then define interfaces between zones (in this case 1-1 abutting).

Regarding the second approach I have a couple of questions. To make the code work I did the following:
  
```
   call cgp_open_f(fname,CG_MODE_WRITE,cgfile,ier)
   if (ier .ne. CG_OK) call cgp_error_exit_f
   basename = 'Base'
   call cg_base_write_f(cgfile,basename,icelldim,iphysdim,cgbase,ier)
   nb = my_rank +  1 # processor block number
   do i=1,np #Number of processess
      isize(1,1) = NNT_bl(i) #local  size of nodes processor i
      isize(1,2) = NTE_bl(i) #local size of elemetns of processor i
      isize(1,3) = 0
      write(zonename,'(a,I4.4)') 'Zone',i
      call cg_zone_write_f(cgfile,cgbase,zonename,isize,Unstructured,cgzone(i),ier)
      call cgp_coord_write_f(cgfile,cgbase,cgzone(i),RealDouble,'CoordinateX',cgcorid(1,i),ier)
      .....
   enddo
   call cgp_coord_write_data_f(cgfile,cgbase,cgzone(nb),cgcorid(1,nb),&
           nnodes_start,nnodes_end,x,ierr)
   ....

```
The above code is executed by each processor, so each processor does np calls to `cg_zone_write`. In case I call the `cg_zone write` only for the processor zone e.g.
```

       nb = my_rank +  1 # processor block number
      isize(1,1) = NNT #local  size of nodes processor i
      isize(1,2) = NTE #local size of elemetns of processor i
      isize(1,3) = 0
      write(zonename,'(a,I4.4)') 'Zone',nb
      call cg_zone_write_f(cgfile,cgbase,zonename,isize,Unstructured,cgzone(nb),ier)
      call cgp_coord_write_f(cgfile,cgbase,cgzone(nb),RealDouble,'CoordinateX',cgcorid(1,nb),ier)
      .....
      call cgp_coord_write_data_f(cgfile,cgbase,cgzone(nb),cgcorid(1,nb),&
               nnodes_start,nnodes_end,x,ierr)
```
It doesn't work. The same holds for all metadata (`cg_sections_write,cg_sol_write, etc`). Is this the right way to do it? 

The final question regards the intrefaces between the zones. I invoke `cg_conn_write_f `which is  a serial subroutine. Of course each processors interfaces with different domains with different number of connections. 

I tried to invoke it like:

```
do i=1, neighbouring_zones_of_my_rank\n
      find connectivity
     call cg_conn_write_f
enddo

call cgp_close
```
I tried it initially with two processes and it seemd to work. However,  I open the cgns file with _vitables_ to see the data and found out that the ZoneGridConnectivity is defined for only one of the two zones. For three or more processes it hangs.

Has anyone some tip about this?

Regards,
George

**Comment by papis84** [comment]
UPDATE
--------
I think that  the basic problem in the multizone case is that when cg_routines are called by all the processors it writes only for the 0th rank in the file.
 
I suppose that I am doing something wrong because if all the processors need to know all the metadata (number of BC's for each zone, connectivities, etc) then there will be surely a deadlock as the number of cpus increases.

Any ideas?

**Reply by MicK7** [reply]
The PCGNS API seems to be fitted for the case where all the processors are writing into the same array.
If each processors need to write to separate zones maybe it is better to write separate files

**Reply by brtnfld** [reply]
I'm not sure this will help, but here is one example of using CGNS with Metis.
 https://github.com/obrienadam/Partition-CGNS-Mesh-Example

I would advise you to avoid adding number-of-rank dependencies into the CGNS file. Odds are, the I/O performance will get worse as you increase the number of ranks, mainly due to the increasing amount of metadata generated as you increase the number of ranks. For parallel I/O, you want to minimize the metadata overhead as much as possible.

**Reply by papis84** [reply]
Thank a lot for your advise ! I was not sure whether I had it right or not!

So basically our approach comes down to :
1) each process writes each own h5 file
2) create a post process script using cg_ routines that read all the files and assembles the output. 

This way we can also get rid of the zone connectivity overhead since we can transform the local cpu numbering (nodes, elements, faces) to the  global one.
